File: C++/234_Palindrome_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == NULL) {
            return true;
        }

        ListNode* mid = find_middle(head);
        ListNode* node = reverse(mid->next);
        mid->next = NULL;

        while (node != NULL && head->val == node->val) {
            head = head->next;
            node = node->next;
        }
        return node == NULL;
    }
private:
    ListNode* get_middle(ListNode *head) {
        if (head == NULL) {
            return NULL;
        }

        ListNode* fast = head->next;
        ListNode* slow = head;

        while (fast != NULL && fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
        }

        return slow;
    }

    ListNode* reverse(ListNode* head) {
        ListNode* prev = NULL;

        while (head != NULL) {
            ListNode* tmp = head->next;
            head->next = prev;
            prev = head;
            head = tmp;
        }

        return prev;
    }
};

File: Algorithms/JavaScript/Easy/234_Palindrome_Linked_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

// O(n) Time Complexity, O(1) Space Complexity

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if (head === null) {
        return true
    }

    var mid = getMiddle(head)
    var left = head
    var right = mid.next
    mid.next = null

    right = reverse(right)

    while (right !== null) {
        if (left.val !== right.val) {
            return false
        }
        left = left.next
        right = right.next
    }
    return true
};

var getMiddle = function(head) {
    if (head === null) {
        return null
    }

    var slow = head
    var fast = head.next

    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }

    return slow
}

var reverse = function(head) {
    if (head === null) {
        return null
    }

    var previous = null
    while (head !== null) {
        var tmp = head.next
        head.next = previous
        previous = head
        head = tmp
    }

    return previous
}
